influx-reports/influx.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/env python3
+""" Get all measurements for an influx job, and
+outputs a report of uptime for maas_ components.
+Inputs:
+    - env.INFLUX_IP (mandatory)
+    - env.INFLUX_PORT (mandatory)
+    - env.BUILD_TAG (mandatory)
+    - ymlreport <ymlreportfile> (optional)
+    - subunitreport <subunitreport> (optional)
+Outputs:
+    - stdout
+    - yaml file with uptime (when ymlreport is used)
+    - subunit binary file with passing/failure criteria
+    (when subunitreport is used)
+"""
+
+import argparse
+from collections import defaultdict
+import datetime
+import logging
+import os
+import sys
+import yaml
+
+import dateutil.parser
+from influxdb import InfluxDBClient
+from subunit.v2 import StreamResultToBytes
+
+
+class InfluxTimestampParseException(Exception):
+    pass
+
+
+class SubunitContext():
+    """Context manager for writing subunit results."""
+
+    def __init__(self, output_path):
+        self.output_path = output_path
+
+    def __enter__(self):
+        self.output_stream = open(self.output_path, 'wb')
+        self.output = StreamResultToBytes(self.output_stream)
+        self.output.startTestRun()
+        return self.output
+
+    def __exit__(self, *args, **kwargs):
+        self.output.stopTestRun()
+        self.output_stream.close()
+
+    def status(self, *args, **kwargs):
+        self.output.status(*args, **kwargs)
+
+
+def return_time(client, query, delta_seconds=0):
+    """ From an InfluxDB query, fetch
+    the first point time, and return a
+    python time object. Shift it from
+    a few seconds (delta_seconds) if necessary.
+    """
+    timestamp_query = client.query(query)
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(
+            fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as alle:
+        raise InfluxTimestampParseException(
+            "Error parsing a timestamp from influx: {}".format(alle))
+    return fpt
+
+
+def calculate_measurement_uptime(client, measurement, job_reference,
+                                 measurement_period_seconds,
+                                 resolution=60):
```

Comment:
Points are happening not one per 60 seconds, but one per 30 seconds with jitter.
Having a large enough resolution (2*sampling) should be good enough.

Minutes can be used as unit, but we used seconds everywhere else in our variables.
It makes sense to re-use the same unit.